### PR TITLE
Feature/cache results

### PR DESCRIPTION
**Pull request by campos20**
Minimum impact in fixing rankings

### CODE REVIEW

**Comment by campos20** [comment]
Early PR open to discuss. Each event should cache up to about 700kb (years excluded). I think we could also dismiss the 1000 results for good.

**Comment by campos20** [comment]
We can also dismiss the hash need and cache it in a table similar to

key, rows (json), cache_time

and retrieve it. Cache would be done in the database.

**Comment by campos20** [comment]
Should we display the date when results were cached?

File: WcaOnRails/db/migrate/20220706232200_create_cached_results.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateCachedResults < ActiveRecord::Migration[6.1]
+  def change
+    create_table(:cached_results, id: false) do |t|
+      t.string :key_params
+      t.json :payload
+      t.timestamps
+    end
+    execute "alter table cached_results add primary key (key_params);"
```
Comment (gregorbg) [inline WcaOnRails/db/migrate/20220706232200_create_cached_results.rb:10]: No need for this, see my other comment above.

File: WcaOnRails/db/migrate/20220706232200_create_cached_results.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateCachedResults < ActiveRecord::Migration[6.1]
+  def change
+    create_table(:cached_results, id: false) do |t|
+      t.string :key_params
```
Comment (gregorbg) [inline WcaOnRails/db/migrate/20220706232200_create_cached_results.rb:6]: `t.string :key_params, primary_key: true`

File: WcaOnRails/app/controllers/results_controller.rb [hunk 3]
```diff
@@ -141,14 +128,23 @@ def rankings
...
+
     else
       flash[:danger] = t(".unknown_show")
       redirect_to rankings_path
     end
+
+    @rows = ActiveRecord::Base.connection.exec_query(@query)
+
+    # For safety, we delete possible duplicated (does active redord have on duplicated update or ignore?)
+    safe_duplate_query = "delete from cached_results where key_params = '#{cached_key}'"
```
Comment (gregorbg) [inline WcaOnRails/app/controllers/results_controller.rb:143]: Typo: `duplate` --> `duplicate`.
Also, `CachedResult.delete_by(key_params: cached_key)`

**Review by gregorbg — COMMENTED** [review]
Thanks, but why are you still disallowing `year`? I remember our long video call, but that discussion was because `year` broke the new table structure that you had created.
Now, you're caching the **old** existing table structure, which can easily support `year`. Why drop it?

Also, wonderful idea about just clearing the cache with CAD. I like! :D

There's a part in `views/results/rankings.html.erb` which GitHub doesn't let me comment on. It has a block with `@skip_robot_indexing` that can be entirely deleted now.

**Comment by gregorbg** [comment]
On second thought, I am a bit concerned whether this will solve our problem of bots killing the server.
If Chinese crawlers decide to try all Results permutations at once, it will still kill/overload the DB during initial computation. The fact that we cache stuff _after_ the initial load doesn't change the problem of initial load. Thoughts?

**Comment by campos20** [comment]
> On second thought, I am a bit concerned whether this will solve our problem of bots killing the server. If Chinese crawlers decide to try all Results permutations at once, it will still kill/overload the DB during initial computation. The fact that we cache stuff _after_ the initial load doesn't change the problem of initial load. Thoughts?

We can put AWS lambdas cron to run accessing #GET...?params..., one after another, just after CAD (or scheduled independently each 12h so we don't impact developing code). If we put in CAD, we will have to make sure the project has AWS SDK and that the associated IAM role has trigger lambda permissions.

File: docker-compose-local-db.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+version: "3"
```
[the remark this answers is not in the document]
Comment (campos20) [reply docker-compose-local-db.yml:1]: I'll open, but I have already a local WCA database setup and I'd like to run this with docker. Setting up a second WCA db is a waste of effort.

File: WcaOnRails/lib/auxiliary_data_computation.rb [hunk 5]
```diff
@@ -113,4 +115,9 @@ def self.delete_php_cache
     cache_files = Dir.glob(Rails.root.join("../webroot/results/generated/cache/*.cache"))
     FileUtils.rm cache_files
   end
+
+  def self.delete_cached_results
+    # By deleting existing results, they can be cached again on the first execution
+    ActiveRecord::Base.connection.execute "delete from cached_results where 1=1"
```
[the remark this answers is not in the document]
Comment (campos20) [reply WcaOnRails/lib/auxiliary_data_computation.rb]: Fixed

File: WcaOnRails/db/migrate/20220706232200_create_cached_results.rb (hunk 2 above)
Comment (campos20) [reply WcaOnRails/db/migrate/20220706232200_create_cached_results.rb]: Fixed.

File: WcaOnRails/app/controllers/results_controller.rb (hunk 3 above)
Comment (campos20) [reply WcaOnRails/app/controllers/results_controller.rb:151]: Fixed

File: docker-compose-local-db.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (campos20) [reply docker-compose-local-db.yml]: Robot indexing handled

File: WcaOnRails/db/migrate/20220706232200_create_cached_results.rb (hunk 1 above)
Comment (campos20) [reply WcaOnRails/db/migrate/20220706232200_create_cached_results.rb]: Fixed

**Comment by gregorbg** [comment]
Pre-caching the results after CAD is run is too much of a rabbit hole because of AWS SDK etc.

I think if we remove the 1000 option and stick with 100 for now, it will be fine. Can you change the code such that the 1000 Results option is _always_ greyed out with a gentle "WST is working on a more permanent solution" kind of tooltip?

**Comment by campos20** [comment]
> Pre-caching the results after CAD is run is too much of a rabbit hole because of AWS SDK etc.
> 
> I think if we remove the 1000 option and stick with 100 for now, it will be fine. Can you change the code such that the 1000 Results option is _always_ greyed out with a gentle "WST is working on a more permanent solution" kind of tooltip?

Done

image

**Comment by gregorbg** [comment]
Please be more specific about "Years will cause us issues". Again, same discussion as above. Why are you so adamant about years even in the old inefficient table structure?

**Comment by campos20** [comment]
> Please be more specific about "Years will cause us issues". Again, same discussion as above. Why are you so adamant about years even in the old inefficient table structure?

2 Type
[single, average]

43 Years
[all years, 1982,2003-2022 only, 1982,2003-2022 until]

3 Gender
[All, male, female]

3 Show
[100 Person, 100 Results, By Region] 1000 is currently disabled

17 Events
215 regions

Total
2828970 keys

Each with about 45kb. Total up to 121gb
Each year adds to new keys ("only" and "until"), 131580 new keys, or up to 5gb.

**Review by gregorbg — APPROVED** [review]
Thanks, this is an explanation that I can live with 😄

**Pull request merged**